lib-rpc/ocamlformat_rpc_lib.ml

```diff
@@ -9,194 +9,199 @@
+      | `Version v -> List [Atom "Version"; Atom v] | _ -> assert false
 
-    let query command t =
-      Command.output t.output command ;
-      Command.read_input t.input
-
-    let halt t =
-      match
-        Command.output t.output `Halt ;
-        close_in t.input ;
-        close_out t.output
-      with
-      | exception _ -> Error (`Msg "failing to close connection to server")
-      | () -> Ok ()
-
-    let config c t =
-      match query (`Config c) t with
-      | `Config _ -> Ok ()
-      | `Error msg -> Error (`Msg msg)
-      | _ -> Error (`Msg "failing to set configuration: unknown error")
+    let output oc t = IO.write oc [to_sexp t]
+  end
 
-    let format x t =
-      match query (`Format x) t with
-      | `Format x -> Ok x
-      | `Error msg -> Error (`Msg msg)
-      | _ -> Error (`Msg "failing to format input: unknown error")
+  module V1 :
+    V
+      with type Command.t =
+        [ `Halt
+        | `Unknown
+        | `Error of string
+        | `Config of (string * string) list
+        | `Format of string ] = struct
+    module Command = struct
+      type t =
+        [ `Halt
+        | `Unknown
+        | `Error of string
+        | `Config of (string * string) list
+        | `Format of string ]
+
+      let read_input ic =
+        let open Csexp in
+        let open IO in
+        read ic
+        >>= function
+        | None -> return `Halt
+        | Some (List [Atom "Format"; Atom x]) -> return (`Format x)
+        | Some (List [Atom "Config"; List l]) ->
+            let c =
+              List.fold_left
+                (fun acc -> function
+                  | List [Atom name; Atom value] -> (name, value) :: acc
+                  | _ -> acc )
+                [] l
+              |> List.rev
+            in
+            return (`Config c)
+        | Some (List [Atom "Error"; Atom x]) -> return (`Error x)
+        | Some (Atom "Halt") -> return `Halt
+        | Some _ -> return `Unknown
+
+      let to_sexp =
+        let open Csexp in
+        function
+        | `Format x -> List [Atom "Format"; Atom x]
+        | `Config c ->
+            let l =
+              List.map (fun (name, value) -> List [Atom name; Atom value]) c
+            in
+            List [Atom "Config"; List l]
+        | `Error x -> List [Atom "Error"; Atom x]
+        | `Halt -> Atom "Halt"
+        | _ -> assert false
+
+      let output oc t = IO.write oc [to_sexp t]
+    end
+
+    module Client = struct
+      type t = {pid: int; input: IO.ic; output: IO.oc}
+
+      type cmd = Command.t
+
+      let pid t = t.pid
+
+      let mk ~pid input output = {pid; input; output}
+
+      let query command t =
+        let open IO in
+        Command.output t.output command
+        >>= fun () -> Command.read_input t.input
+
+      let halt t =
+        let open IO in
+        match Command.output t.output `Halt with
+        | exception _ ->
+            return (Error (`Msg "failing to close connection to server"))
```

Comment:
This error might be ignored by the caller but the `close` must happen anyway.

Reply:
Taken care of by the last change